File: usaco/training/section_5.3/bigbrn.cpp
/*
ID: huangch7
LANG: C++
TASK: bigbrn
*/
#include <iostream>
#include <fstream>
#include <vector>
#include <algorithm>

using namespace std;

int main() {
    ifstream fin("bigbrn.in");
    ofstream fout("bigbrn.out");

    int N, T;
    fin >> N >> T;

    vector<vector<int>> dp(N, vector<int>(N, 1)); // Initialize with 1s
    vector<pair<int, int>> barns;
    
    for (int i = 0; i < T; ++i) {
        int x, y;
        fin >> x >> y;
        barns.push_back({x - 1, y - 1}); // Adjust indices to 0-based
        dp[x - 1][y - 1] = 0;
    }

    int best = 0;
    for (int i = 0; i < N; ++i) {
        best = max(dp[i][0], best);
        best = max(dp[0][i], best);
    }

    for (int i = 1; i < N; ++i) {
        for (int j = 1; j < N; ++j) {
            if (dp[i][j] == 0) continue;
            if (dp[i - 1][j] == 0) continue;
            if (dp[i][j - 1] == 0) continue;
            
            int n = min(dp[i][j - 1], dp[i - 1][j]);
            if (dp[i - n][j - n]) {
                dp[i][j] = max(n + 1, dp[i][j]);
            }
            else{
              dp[i][j] = max(n, dp[i][j]);
            }

            if (best < dp[i][j]) {
                best = dp[i][j];
            }
        }
    }

    fout << best << endl;

    fin.close();
    fout.close();

    return 0;
}

File: usaco/training/section_5.3/bigbrn.py
"""
ID: huangch7
LANG: PYTHON3
TASK: bigbrn
"""
import math

task_name = "bigbrn"
INF = 100000000


def get_list_from_line(dtype=int):
    return [dtype(x) for x in fin.readline().strip().split()]


def get_strs_from_line():
    return fin.readline().strip().split()



fin = open(f'{task_name}.in', 'r')
fout = open(f'{task_name}.out', 'w')

N, T = get_list_from_line()
barns = []
dp = [[1] * N for _ in range(N)]
for _ in range(T):
    x, y = get_list_from_line()
    barns.append((x, y))
    dp[x - 1][y - 1] = 0

best = 0
for i in range(N):
    best = max(dp[i][0], best)
    best = max(dp[0][i], best)
for i in range(1, N):
    for j in range(1, N):
        if dp[i][j] == 0:
            continue
        if dp[i - 1][j] == 0:
            continue
        if dp[i][j - 1] == 0:
            continue
        n = min(dp[i][j - 1], dp[i - 1][j])
#         while n > 0:
#             if dp[i - n][j - n]:
#                 dp[i][j] = max(n + 1, dp[i][j])
#             n -= 1
        if dp[i - n][j - n]:
            dp[i][j] = max(n + 1, dp[i][j])
        else:
            dp[i][j] = max(n, dp[i][j])
            
        if best < dp[i][j]:
            best = dp[i][j]
#     print(dp[i])
fout.write(f"{best}\n")
fout.close()
